dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,245 @@
+        # mem_chunk_id must be unique among all workers and callbacks in the pool,
+        # used to identify shared memory chunks in the communication between processes
+        self.mem_chunk_id = mem_chunk_id
+        self.shm_chunk = shared_mem.SharedMem.allocate(capacity)
+        self.capacity = capacity
+
+    def resize(self, new_capacity=None):
+        capacity = self.capacity
+        new_capacity = new_capacity or 2 * capacity
+        self.shm_chunk.resize(new_capacity, trunc=True)
+        self.capacity = new_capacity
+
+    def close(self):
+        # Shared memory handle and mapping will be freed automatically when shm pybind wrapper gets
+        # garbage collected anyway, but you can be so nice to free them as soon as you know
+        # you won't use them
+        self.shm_chunk.close()
+
+
+def _to_numpy(sample):
+    if isinstance(sample, np.ndarray):
+        return sample
+    if types._is_mxnet_array(sample):
+        return sample.asnumpy()
+    if types._is_torch_tensor(sample):
+        if sample.device.type != 'cpu':
+            raise TypeError("GPU tensors are not supported")
+        return sample.numpy()
+    elif isinstance(sample, tensors.TensorCPU):
+        return np.array(sample)
+    raise TypeError(
+        "Unsupported callback return type. Expected numpy array, pytorch or mxnet cpu tensors, "
+        "or list or tuple of them.")
+
+
+def _apply_to_sample(func, sample, *args):
+    """Apply to a sample traversing the nesting of the data (tuple/list).
+
+    Parameters
+    ----------
+    func : callable
+        Function to be applied to every sample data object
+    sample : sample object or any nesting of those in tuple/list
+        Representation of sample
+    """
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(_apply_to_sample(func, part, *args) for part in sample)
+    else:
+        # we unpacked all nesting levels, now is actual data:
+        return func(sample, *args)
+
+
+class SharedBatchWriter:
+    """SharedBatchWriter can serialize and write batch into given shared
+    memory chunk (``mem_batch``).
+    """
+
+    def __init__(self, mem_batch: SharedMemChunk, batch):
+        import_numpy()
+        self.mem_batch = mem_batch
+        self.data_size = 0
+        self.meta_data_size = 0
+        self._write_batch(batch)
+
+    def _prepare_samples_meta(self, indexed_samples):
+        """Calculate metadata and total size of data to be serialized"""
+        data_size = 0
+
+        def make_meta(np_array):
+            nonlocal data_size
+            offset = data_size
+            data_size += np_array.nbytes
+            return SampleMeta(offset, np_array.shape, np_array.dtype, np_array.nbytes)
+
+        meta = []
+        for idx, sample in indexed_samples:
+            meta.append((idx, _apply_to_sample(make_meta, sample)))
+        return meta, data_size
+
+    def _add_array_to_batch(self, np_array, memview):
```

Comment:
```suggestion
    def _add_array_to_batch(self, np_array, memview, offset):
```
Calculating offset _again_ while we already have it in `meta` is error prone. We should have a single source of truth.

Reply:
Nesting was painful. Will try to do something.